shared/Microsoft.Extensions.WebEncoders.Sources/WebEncoders.cs

```diff
@@ -252,137 +389,774 @@ public static string Base64UrlEncode(byte[] input, int offset, int count)
+
+            var numWholeOrPartialInputBlocks = (dataLength + 2) / 3;
+            return numWholeOrPartialInputBlocks * 4;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int GetBufferSizeRequiredToBase64Encode(int dataLength, out int numPaddingChars)
+        {
+            // Shortcut for Guid and other 16 byte data
+            if (dataLength == 16)
+            {
+                numPaddingChars = 2;
+                return 24;
+            }
+
+            numPaddingChars = GetNumBase64PaddingCharsAddedByEncode(dataLength);
+            return GetBufferSizeRequiredToBase64Encode(dataLength);
         }
 
-        private static int GetNumBase64PaddingCharsInString(string str)
+        private static int GetNumBase64PaddingCharsAddedByEncode(int dataLength)
         {
-            // Assumption: input contains a well-formed base64 string with no whitespace.
+            // Calculation is:
+            // switch (dataLength % 3)
+            // 0 -> 0
+            // 1 -> 2
+            // 2 -> 1
+
+            return dataLength % 3 == 0 ? 0 : 3 - dataLength % 3;
+        }
 
-            // base64 guaranteed have 0 - 2 padding characters.
-            if (str[str.Length - 1] == '=')
+        private static void ThrowInvalidArguments(object input, int offset, int count, char[] buffer = null, int bufferOffset = 0, ExceptionArgument bufferName = ExceptionArgument.buffer, bool validateBuffer = false)
+        {
+            throw GetInvalidArgumentsException();
+
+            Exception GetInvalidArgumentsException()
             {
-                if (str[str.Length - 2] == '=')
+                if (input == null)
                 {
-                    return 2;
+                    return ThrowHelper.GetArgumentNullException(ExceptionArgument.input);
                 }
-                return 1;
+
+                if (validateBuffer && buffer == null)
+                {
+                    return ThrowHelper.GetArgumentNullException(bufferName);
+                }
+
+                if (offset < 0)
+                {
+                    return ThrowHelper.GetArgumentOutOfRangeException(ExceptionArgument.offset);
+                }
+
+                if (count < 0)
+                {
+                    return ThrowHelper.GetArgumentOutOfRangeException(ExceptionArgument.count);
+                }
+
+                if (bufferOffset < 0)
+                {
+                    return ThrowHelper.GetArgumentOutOfRangeException(ExceptionArgument.bufferOffset);
+                }
+
+                return ThrowHelper.GetInvalidCountOffsetOrLengthException(ExceptionArgument.count, ExceptionArgument.offset, ExceptionArgument.input);
             }
-            return 0;
         }
 
-        private static int GetNumBase64PaddingCharsToAddForDecode(int inputLength)
+        // TODO: replace IntPtr and (int*) with nuint once available
```

Comment:
`nuint` isn't avaialbe, so it is a bit messy.

The other workaround we be defining an `using alias`, depending on an `#if` for the bitness. On this linked file I didn't know how to do this, so I used the `IntPtr`-approach.

The TODO comment is to remind me for the change once available.